### DISCUSSION TITLE
Export circular relationship

### DISCUSSION THREAD

**Discussion by scajal** [pxlrbt/filament-excel#167] [Q&A] [answered]
I have a `users` table that contains a `referred_by` column that points to the same table. 
For viewing the referrals, I have a custom relation page with a table. 

I need to export the table in this page, but the Excel is automatically filtered by the owner record:
The query is correctly created, but a `WHERE user_id IN (<owner record id>)` is added. I'd need this where to be removed.

Is there any possible way to do that? Or manually set the query I want to use?

I tried using `modifyQueryUsing`, but at that point, the `where` clause is not added yet.

**Comment by pxlrbt** [comment]
I am not sure why the export would add a where clause? It should simply take the table or model query and use that one 🤔

**Comment by scajal** [comment, marked as answer]
@pxlrbt the `where` is being added here: https://github.com/pxlrbt/filament-excel/blob/c4147a3c91ac9f9b0aca786174f777a593ac3171/src/Exports/ExcelExport.php#L278
I suppose that's because it's inside a relationship.

Anyway, I solved it by overriding the `export` method of the `pxlrbt\FilamentExcel\Exports\ExcelExport` class and removing the `where` before making the export (implementing a custom closure):

```php
public function export()
{
    if ($this->modifyQueryBeforeExportUsing) {
        app()->call($this->modifyQueryBeforeExportUsing->getClosure(), [
            'query' => $this->getQuery(),
            'livewire' => $this->getLivewire(),
        ]);
    }

    return parent::export();
}
```

Then, I'd do:

```php
\App\Filament\Pages\Actions\ExcelExport::make()
  ->modifyQueryBeforeExportUsing(static function (Builder $query): void {
      // Remove the owner record filter from the 
      // query, to prevent wrong exports.
      $query->getQuery()->wheres = array_filter(
          $query->getQuery()->wheres,
          static function (array $where): bool {
              return !array_key_exists('column', $where)
                  || $where['column'] !== 'users.id';
          }
      );
  })
```

Not so pretty solution I think, but works.

**Reply by pxlrbt** [reply]
Ah. Actually that `->where()` is only meant for bulk actions and the selected entries. Not sure when other actions also added `$recordIds` 🤔

But glad you found a solution.
